pkg/rpc/nodedialer/nodedialer.go

```diff
@@ -217,35 +218,36 @@ func (n *Dialer) ConnHealth(nodeID roachpb.NodeID, class rpc.ConnectionClass) er
 // a connection attempt in the background if it doesn't and always return
 // immediately. It is only used today by DistSQL and it should probably be
 // removed and moved into that code. Also, as of #99191, we have stateful
-// circuit breakers that probe in the background and so whatever exactly it
-// is the caller really wants can likely be achieved by more direct means.
-func (n *Dialer) ConnHealthTryDial(nodeID roachpb.NodeID, class rpc.ConnectionClass) error {
-	err := n.ConnHealth(nodeID, class)
-	if err == nil {
-		return err
+// circuit breakers that probe in the background and so whatever exactly it is
+// the caller really wants can likely be achieved by more direct means.
+func (n *Dialer) ConnHealthTryDial(
+	id base.SQLInstanceID, addr string, locality roachpb.Locality,
+) error {
+	if n == nil {
+		return errors.New("no node dialer configured")
 	}
-	addr, locality, err := n.resolver(nodeID)
-	if err != nil {
-		return err
+	if err := n.rpcContext.ConnHealth(
+		addr, roachpb.NodeID(id), rpc.DefaultClass); err == nil {
+		return nil
 	}
 	// NB: This will always return `ErrNotHeartbeated` since the heartbeat will
 	// not be done by the time `Health` is called since GRPCDialNode is async.
-	return n.rpcContext.GRPCDialNode(addr.String(), nodeID, locality, class).Health()
+	return n.rpcContext.GRPCDialInstance(addr, id, locality, rpc.DefaultClass).Health()
 }
 
-// ConnHealthTryDialInstance returns nil if we have an open connection of the
-// rpc.DefaultClass to the given sqlinstance that succeeded on its most recent
-// heartbeat. If no healthy connection is found, it will attempt to dial the
-// instance.
-func (n *Dialer) ConnHealthTryDialInstance(id base.SQLInstanceID, addr string) error {
-	if n == nil {
+// ConnHealthTryDialWithResolve returns nil if we have an open connection of the
+// request class to the given node that succeeded on its most recent heartbeat.
+// If no healthy connection is found, it will attempt to dial the node.
+func (n *Dialer) ConnHealthTryDialWithResolve(id base.SQLInstanceID) error {
+	if n == nil || n.resolver == nil {
 		return errors.New("no node dialer configured")
 	}
-	if err := n.rpcContext.ConnHealth(
-		addr, roachpb.NodeID(id), rpc.DefaultClass); err == nil {
-		return nil
+	addr, locality, err := n.resolver(roachpb.NodeID(id))
+	if err != nil {
+		return err
 	}
-	return n.rpcContext.GRPCDialPod(addr, id, roachpb.Locality{}, rpc.DefaultClass).Health()
```

Comment:
I wonder if passing this empty locality had any implications earlier.

Reply:
The locality only matters for metrics. Ideally it should pass the `remoteLocality`, which is the Locality of the node it is talking to. I'm not sure if all the callers have the locality. If this connection is only opened to check health and then closed, it won't matter in any meaningful way, but is still better to fix.

Reply:
Thanks for answering this. Yes, I've fixed it in this PR.